### PR TITLE
Check and resolve conflict in RowsetUpdateStates::apply

### PR DESCRIPTION
**Pull request by sevev**
Check and resolve conflict in RowsetUpdateStates::apply if conflict exists
1. If the row has been upsert, re-read this row's value
2. If the row has been deleted, set this row's value to its default value

### CODE REVIEW

**Review by chaoyli — APPROVED** [review]

File: be/src/storage/rowset_update_state.cpp [hunk 1]
```diff
@@ -232,20 +232,110 @@ Status RowsetUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset*
...
+        index.get(pks, &new_rss_rowids[i]);
+    }
+    int64_t t_get_rowids = MonotonicMillis();
+
+    for (uint32_t i = 0; i < num_segments; ++i) {
+        std::vector<std::unique_ptr<vectorized::Column>> new_write_columns;
+        new_write_columns.resize(_partial_update_states[i].write_columns.size());
+        for (uint32_t j = 0; j < new_write_columns.size(); ++j) {
+            new_write_columns[j] = _partial_update_states[i].write_columns[j]->clone_empty();
+        }
```
Comment (decster) [inline be/src/storage/rowset_update_state.cpp:282]: Only create this when there is conflict

File: be/src/storage/rowset_update_state.cpp [hunk 2]
```diff
@@ -232,20 +232,110 @@ Status RowsetUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset*
...
+            }
+        }
+        if (has_conflict) {
+            size_t num_default = 0;
+            std::map<uint32_t, std::vector<uint32_t>> rowids_by_rssid;
+            std::vector<uint32_t> read_idxes;
+            plan_read_by_rssid(conflict_rowids, &num_default, &rowids_by_rssid, &read_idxes);
+            DCHECK_EQ(conflict_idxes.size(), read_idxes.size());
+            RETURN_IF_ERROR(tablet->updates()->get_column_values(read_column_ids, num_default > 0, rowids_by_rssid,
+                                                                 &_partial_update_states[i].write_columns));
```
Comment (decster) [inline be/src/storage/rowset_update_state.cpp:313]: the number of conflicts is usually very small, could we avoid copying the entire column? just set the rows that need to be updated.

File: be/src/storage/rowset_update_state.cpp [hunk 3]
```diff
@@ -232,20 +232,110 @@ Status RowsetUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset*
...
+                                                             idxes.data(), 0, idxes.size());
+            }
+
+            _partial_update_states[i].write_columns.swap(new_write_columns);
+        }
+    }
+
+    int64_t t_resolve_conflict = MonotonicMillis();
+    LOG(INFO) << Substitute(
+            "check partial rowset conflict tablet:$0 rowset:$1 #column:$2 getrowid:$3ms #resolve_conflict $4ms",
```
Comment (decster) [inline be/src/storage/rowset_update_state.cpp:330]: also log number of conflict/updated rows rows

File: be/src/storage/rowset_update_state.cpp [hunk 4]
```diff
@@ -232,20 +232,110 @@ Status RowsetUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset*
...
+            _partial_update_states[i].write_columns[col_idx]->append_selective(*read_columns[col_idx], idxes.data(), 0,
                                                                                idxes.size());
         }
     }
 
     return Status::OK();
 }
 
-Status RowsetUpdateState::apply(Tablet* tablet, Rowset* rowset, uint32_t rowset_id, const PrimaryIndex& index) {
+Status RowsetUpdateState::_check_conflict(Tablet* tablet, Rowset* rowset, uint32_t rowset_id,
```
Comment (decster) [inline be/src/storage/rowset_update_state.cpp:248]: this method not only check conflict, but also resolve the conflict, so maybe change to _check_and_resolve_conflict ?

**Review by decster — CHANGES_REQUESTED** [review]

**Comment by sevev** [comment]
duplicate of https://github.com/StarRocks/starrocks/pull/2919

**Pull request closed without merging**